File: backend/agent/skills.py
"""Skill discovery and parsing for Warden.

A skill is a Markdown instruction set in `<dir>/SKILL.md` that the agent
can load on demand. Optional YAML frontmatter (name, description) is
parsed; the body is the actual instructions.

Discovery paths (low to high priority — later wins on duplicate name):
- ~/.codex/skills/   (legacy)
- ~/.agents/skills/
- ~/.claude/skills/   (compat)
- ~/.warden/skills/
- <cwd>/skills/      (legacy bare project)
- <cwd>/.claude/skills/   (compat)
- <cwd>/.warden/skills/   (canonical project)
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from pathlib import Path

log = logging.getLogger(__name__)

MAX_SKILL_BYTES = 64 * 1024  # hard cap; beyond this, the body is truncated

# directory name must match this
_NAME_RE = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")

# frontmatter at top of file: ---\nkey: value\n...\n---\nbody
_FRONTMATTER_RE = re.compile(r"\A---\s*\n(.*?)\n---\s*\n(.*)", re.DOTALL)


@dataclass
class Skill:
    name: str
    description: str
    content: str  # body, frontmatter stripped
    location: str  # absolute path to SKILL.md
    directory: str  # absolute path to skill dir


def _skill_roots() -> list[Path]:
    home = Path.home()
    return [
        home / ".codex" / "skills",
        home / ".agents" / "skills",
        home / ".claude" / "skills",
        home / ".warden" / "skills",
        Path.cwd() / "skills",
        Path.cwd() / ".claude" / "skills",
        Path.cwd() / ".warden" / "skills",
    ]


def _parse_frontmatter(text: str) -> tuple[dict[str, str], str]:
    """Extract simple `key: value` frontmatter. Returns (meta, body)."""
    m = _FRONTMATTER_RE.match(text)
    if not m:
        return {}, text
    meta: dict[str, str] = {}
    for line in m.group(1).splitlines():
        line = line.rstrip()
        if not line or line.startswith("#"):
            continue
        if ":" not in line:
            continue
        key, _, value = line.partition(":")
        key = key.strip().lower()
        value = value.strip().strip('"').strip("'")
        if key and value:
            meta[key] = value
    return meta, m.group(2).lstrip("\n")


def _validate_name(name: str) -> bool:
    return bool(name) and len(name) <= 64 and bool(_NAME_RE.match(name))


def _parse_skill_file(skill_md: Path) -> Skill | None:
    """Parse a single SKILL.md file. Returns None if invalid."""
    try:
        raw = skill_md.read_text(encoding="utf-8", errors="replace")
    except OSError as e:
        log.warning("skill: cannot read %s: %s", skill_md, e)
        return None

    if len(raw.encode("utf-8")) > MAX_SKILL_BYTES:
        raw = raw.encode("utf-8")[:MAX_SKILL_BYTES].decode("utf-8", errors="ignore")
        log.warning("skill: %s truncated to %d bytes", skill_md, MAX_SKILL_BYTES)

    meta, body = _parse_frontmatter(raw)

    dir_name = skill_md.parent.name
    fm_name = meta.get("name", "").strip()
    # frontmatter name wins if present and valid; otherwise directory name
    name = fm_name if _validate_name(fm_name) else dir_name
    if not _validate_name(name):
        log.warning("skill: %s has invalid name %r — skipped", skill_md, name)
        return None

    if name != dir_name and fm_name:
        log.warning(
            "skill: %s — frontmatter name %r does not match dir %r, using frontmatter",
            skill_md,
            fm_name,
            dir_name,
        )

    description = meta.get("description", "").strip()
    if not description:
        log.warning("skill: %s — missing description in frontmatter", skill_md)

    return Skill(
        name=name,
        description=description,
        content=body.rstrip() + "\n",
        location=str(skill_md.resolve()),
        directory=str(skill_md.parent.resolve()),
    )


def discover_skills() -> list[Skill]:
    """Scan all roots and return skills, deduped by name (highest priority wins)."""
    by_name: dict[str, Skill] = {}
    for root in _skill_roots():
        if not root.is_dir():
            continue
        try:
            children = sorted(root.iterdir(), key=lambda p: p.name.lower())
        except OSError as e:
            log.warning("skill: cannot list %s: %s", root, e)
            continue
        for child in children:
            if not child.is_dir():
                continue
            skill_md = child / "SKILL.md"
            if not skill_md.is_file():
                continue
            parsed = _parse_skill_file(skill_md)
            if parsed is None:
                continue
            # later roots override earlier ones for the same name
            by_name[parsed.name] = parsed
    return sorted(by_name.values(), key=lambda s: s.name)


def find_skill(name: str) -> Skill | None:
    """Find a single skill by name."""
    if not _validate_name(name):
        return None
    for skill in discover_skills():
        if skill.name == name:
            return skill
    return None


def list_siblings(skill: Skill, limit: int = 10) -> list[str]:
    """List sibling files in the skill directory (up to limit)."""
    d = Path(skill.directory)
    out: list[str] = []
    try:
        entries = sorted(d.iterdir(), key=lambda p: p.name.lower())
    except OSError:
        return out
    for entry in entries:
        if entry.name == "SKILL.md":
            continue
        out.append(f"{entry.name}/" if entry.is_dir() else entry.name)
        if len(out) >= limit:
            break
    return out


def format_catalog(skills: list[Skill] | None = None) -> str:
    """Render the available skills as an <available_skills> XML block."""
    items = skills if skills is not None else discover_skills()
    if not items:
        return ""
    lines = [
        "Skills provide specialized instructions and workflows for specific tasks.",
        "Use the skill tool to load a skill when a task matches its description.",
        "<available_skills>",
    ]
    for s in items:
        if not s.description:
            continue
        lines.append("  <skill>")
        lines.append(f"    <name>{s.name}</name>")
        lines.append(f"    <description>{s.description}</description>")
        lines.append("  </skill>")
    lines.append("</available_skills>")
    return "\n".join(lines)


def wrap_skill_content(skill: Skill) -> str:
    """Wrap a skill's body in the <skill_content> payload format."""
    files = list_siblings(skill)
    files_block = "\n".join(files) if files else "(no extra files)"
    return (
        f'<skill_content name="{skill.name}">\n'
        f"# Skill: {skill.name}\n"
        f"\n"
        f"{skill.content.rstrip()}\n"
        f"\n"
        f"Base directory for this skill: {skill.directory}\n"
        f"Relative paths in this skill (e.g., scripts/, references/) are relative to this base directory.\n"
        f"Note: file list is sampled.\n"
        f"\n"
        f"<skill_files>\n"
        f"{files_block}\n"
        f"</skill_files>\n"
        f"</skill_content>"
    )

File: agent/skills/skills.go
package skills

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

const maxSkillBytes = 64 * 1024

var nameRe = regexp.MustCompile(`^[a-z0-9]+(-[a-z0-9]+)*$`)
var frontmatterRe = regexp.MustCompile(`\A---\s*\n(.*?)\n---\s*\n(.*)`)

type Skill struct {
	Name        string
	Description string
	Content     string
	Location    string
	Directory   string
}

func skillRoots() []string {
	home, _ := os.UserHomeDir()
	cwd, _ := os.Getwd()
	return []string{
		filepath.Join(home, ".codex", "skills"),
		filepath.Join(home, ".agents", "skills"),
		filepath.Join(home, ".claude", "skills"),
		filepath.Join(home, ".warden", "skills"),
		filepath.Join(cwd, "skills"),
		filepath.Join(cwd, ".claude", "skills"),
		filepath.Join(cwd, ".warden", "skills"),
	}
}

func parseFrontmatter(text string) (map[string]string, string) {
	m := frontmatterRe.FindStringSubmatch(text)
	if m == nil {
		return nil, text
	}

	meta := make(map[string]string)
	for _, line := range strings.Split(m[1], "\n") {
		line = strings.TrimRight(line, "\r ")
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if !strings.Contains(line, ":") {
			continue
		}
		key, val, _ := strings.Cut(line, ":")
		key = strings.TrimSpace(strings.ToLower(key))
		val = strings.TrimSpace(strings.Trim(strings.Trim(val, `"'`), `"'`))
		if key != "" && val != "" {
			meta[key] = val
		}
	}

	body := strings.TrimLeft(m[2], "\n")
	return meta, body
}

func validateName(name string) bool {
	return name != "" && len(name) <= 64 && nameRe.MatchString(name)
}

func parseSkillFile(skillMd string) *Skill {
	data, err := os.ReadFile(skillMd)
	if err != nil {
		return nil
	}

	text := string(data)
	if len(data) > maxSkillBytes {
		data = data[:maxSkillBytes]
		text = string(data)
	}

	meta, body := parseFrontmatter(text)

	dirName := filepath.Base(filepath.Dir(skillMd))
	fmName := strings.TrimSpace(meta["name"])

	name := dirName
	if validateName(fmName) {
		name = fmName
	}

	if !validateName(name) {
		return nil
	}

	description := strings.TrimSpace(meta["description"])

	content := strings.TrimRight(body, "\r\n") + "\n"

	absPath, _ := filepath.Abs(skillMd)
	absDir, _ := filepath.Abs(filepath.Dir(skillMd))

	return &Skill{
		Name:        name,
		Description: description,
		Content:     content,
		Location:    absPath,
		Directory:   absDir,
	}
}

func DiscoverSkills() []Skill {
	byName := make(map[string]*Skill)

	for _, root := range skillRoots() {
		entries, err := os.ReadDir(root)
		if err != nil {
			continue
		}

		// Sort for deterministic order
		sort.Slice(entries, func(i, j int) bool {
			return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
		})

		for _, entry := range entries {
			if !entry.IsDir() {
				continue
			}
			skillMd := filepath.Join(root, entry.Name(), "SKILL.md")
			if _, err := os.Stat(skillMd); os.IsNotExist(err) {
				continue
			}

			parsed := parseSkillFile(skillMd)
			if parsed == nil {
				continue
			}

			byName[parsed.Name] = parsed
		}
	}

	result := make([]Skill, 0, len(byName))
	for _, s := range byName {
		result = append(result, *s)
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].Name < result[j].Name
	})
	return result
}

func FindSkill(name string) *Skill {
	if !validateName(name) {
		return nil
	}
	for _, s := range DiscoverSkills() {
		if s.Name == name {
			return &s
		}
	}
	return nil
}

func ListSiblings(skill *Skill, limit int) []string {
	var out []string
	entries, err := os.ReadDir(skill.Directory)
	if err != nil {
		return out
	}

	sort.Slice(entries, func(i, j int) bool {
		return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
	})

	for _, e := range entries {
		if e.Name() == "SKILL.md" {
			continue
		}
		if e.IsDir() {
			out = append(out, e.Name()+"/")
		} else {
			out = append(out, e.Name())
		}
		if len(out) >= limit {
			break
		}
	}
	return out
}

func FormatCatalog(skills []Skill) string {
	if len(skills) == 0 {
		return ""
	}
	var lines []string
	lines = append(lines, "Skills provide specialized instructions and workflows for specific tasks.")
	lines = append(lines, "Use the skill tool to load a skill when a task matches its description.")
	lines = append(lines, "<available_skills>")
	for _, s := range skills {
		if s.Description == "" {
			continue
		}
		lines = append(lines, "  <skill>")
		lines = append(lines, "    <name>"+s.Name+"</name>")
		lines = append(lines, "    <description>"+s.Description+"</description>")
		lines = append(lines, "  </skill>")
	}
	lines = append(lines, "</available_skills>")
	return strings.Join(lines, "\n")
}

func WrapSkillContent(skill *Skill) string {
	files := ListSiblings(skill, 10)
	filesBlock := "(no extra files)"
	if len(files) > 0 {
		filesBlock = strings.Join(files, "\n")
	}

	var b strings.Builder
	b.WriteString(`<skill_content name="` + skill.Name + `">`)
	b.WriteString("\n")
	b.WriteString("# Skill: " + skill.Name + "\n")
	b.WriteString("\n")
	b.WriteString(strings.TrimRight(skill.Content, "\n") + "\n")
	b.WriteString("\n")
	b.WriteString("Base directory for this skill: " + skill.Directory + "\n")
	b.WriteString("Relative paths in this skill (e.g., scripts/, references/) are relative to this base directory.\n")
	b.WriteString("Note: file list is sampled.\n")
	b.WriteString("\n")
	b.WriteString("<skill_files>\n")
	b.WriteString(filesBlock + "\n")
	b.WriteString("</skill_files>\n")
	b.WriteString("</skill_content>")

	return b.String()
}
